src/operator/z1.rs

```diff
@@ -1,11 +1,159 @@
 //! z^-1 operator delays its input by one timestamp.
 
-use crate::circuit::{
-    operator_traits::{Operator, StrictOperator, StrictUnaryOperator, UnaryOperator},
-    OwnershipPreference, Scope,
+use crate::{
+    algebra::HasZero,
+    circuit::{
+        operator_traits::{Operator, StrictOperator, StrictUnaryOperator, UnaryOperator},
+        Circuit, ExportId, ExportStream, FeedbackConnector, NodeId, OwnershipPreference, Scope,
+        Stream,
+    },
+    circuit_cache_key,
 };
 use std::{borrow::Cow, mem::swap};
 
+circuit_cache_key!(DelayedId<C, D>(NodeId => Stream<C, D>));
+circuit_cache_key!(NestedDelayedId<C, D>(NodeId => Stream<C, D>));
+
+/// Like [`FeedbackConnector`] but specialized for [`Z1`] feedback operator.
+///
+/// Use this API instead of the low-level [`Circuit::add_feedback`] API to
+/// create feedback loops with `Z1` operator.  In addition to being more
+/// concise, this API takes advantage of [caching](`crate::circuit::cache`).
+pub struct DelayedFeedback<P, D> {
+    feedback: FeedbackConnector<Circuit<P>, D, D, Z1<D>>,
+    output: Stream<Circuit<P>, D>,
+    export: Stream<P, D>,
+}
+
+impl<P, D> DelayedFeedback<P, D>
+where
+    P: Clone + 'static,
+    D: Clone + HasZero + 'static,
+{
+    /// Create a feedback loop with `Z1` operator.  Use [`Self::connect`] to
+    /// close the loop.
+    pub fn new(circuit: &Circuit<P>) -> Self {
+        let (ExportStream { local, export }, feedback) =
+            circuit.add_feedback_with_export(Z1::new(D::zero()));
+        Self {
+            feedback,
+            output: local,
+            export,
+        }
+    }
+
+    /// Output stream of the `Z1` operator.
+    pub fn stream(&self) -> &Stream<Circuit<P>, D> {
+        &self.output
+    }
+
+    /// Connect `input` stream to the input of the `Z1` operator.
+    pub fn connect(self, input: &Stream<Circuit<P>, D>) {
+        let Self {
+            feedback,
+            output,
+            export,
+        } = self;
+        let circuit = output.circuit().clone();
+
+        feedback.connect_with_preference(input, OwnershipPreference::STRONGLY_PREFER_OWNED);
+        circuit
+            .cache()
```

Comment:
you always write cache().insert(...), perhaps this should be renamed to just cache(...), or memoize(...).